# Assignment 2

## Abstract


For this assignment we selected one of the Kaggle competitions and used H2O AutoML for getting the best model and its metrics to see where will our kernel lie in the leaderboard if we use this and work ahead with the Kaggle competition and further also try to understand how H2O works.

##### Kaggle competition - Kobe Bryant Shot Selection

The Kaggle competition that we have chosen is - Kobe Bryant Shot Selection - Kobe Bryant marked his retirement from the NBA by scoring 60 points in his final game as a Los Angeles Laker on Wednesday, April 12, 2016. Drafted into the NBA at the age of 17, Kobe earned the sport’s highest accolades throughout his long career.

Using 20 years of data on Kobe's swishes and misses, can you predict which shots will find the bottom of the net? This competition is well suited for practicing classification basics, feature engineering, and time series analysis. Practice got Kobe an eight-figure contract and 5 championship rings. What will it get you?

Link to the competition - https://www.kaggle.com/c/kobe-bryant-shot-selection

##### H2O - AutoML

H2O’s AutoML can be used for automating the machine learning workflow, which includes automatic training and tuning of many models within a user-specified time-limit. Stacked Ensembles – one based on all previously trained models, another one on the best model of each family – will be automatically trained on collections of individual models to produce highly predictive ensemble models which, in most cases, will be the top performing models in the AutoML Leaderboard.


We have applied AutoML to our Kobe Bryant Dataset and tried to check how are the models generated by H2O when we specify it with the target variable and other predictors with it. Thats the main AIM of this assignment. H2O at the backend(server) tries to run various Algorithms that it possesses and for various hyperparameter values will generate various models.

Now, the number of models that H2O generates depends on the amount of time you specify it to run for the dataset loaded.


<span class="girk">`Data & Prediction we are going to make`</span> - Here we are going to predict that weather Kobe Brayants made the shot or not.

`Conclusion` - We, on the way give certain parameters to H2O to get a good model, with a value of Log Loss(which is the deciding factor of our Kaggle competition) which tuns out to be the best - `0.34436`. And thus making us win the competition(if we had participated). Thus we submit this as an assignment to show, that even we students can win a Kaggle competition.


<img src="Images/xrOu1.png" width="300">

## Part 1 - Data cleaning

In [36]:
import numpy as np
import pandas as pd
from xgboost.sklearn import XGBClassifier
from sklearn.model_selection import train_test_split

In [37]:
# Load data and roughly clean it, then sort as game date
df = pd.read_csv("./data.csv")
df.drop(['game_event_id', 'game_id', 'lat', 'lon', 'team_id', 'team_name'], axis=1, inplace=True)
df.sort_values('game_date',  inplace=True)
mask = df['shot_made_flag'].isnull()

### Data Cleaning(Important)

##### This is the most important step that decides our winning factor and getting the best model out of H2O

First of all we will explain the kind of the data and describe about its columns. This data contains the location and circumstances of every field goal attempted by Kobe Bryant took during his 20-year career. Your task is to predict whether the basket went in (shot_made_flag).

We have removed 5000 of the shot_made_flags (represented as missing values in the csv file). These are the test set shots for which you must submit a prediction. You are provided a sample submission file with the correct shot_ids needed for a valid prediction.

To avoid leakage, your method should only train on events that occurred prior to the shot for which you are predicting! Since this is a playground competition with public answers, it's up to you to abide by this rule.

The field names are self explanatory and contain the following attributes:

   * action_type
   * combined_shot_type
   * game_event_id
   * game_id
   * lat
   * loc_x
   * loc_y
   * lon
   * minutes_remaining
   * period
   * playoffs
   * season 
   * seconds_remaining
   * shot_distance
   * <span class="girk">`shot_made_flag`</span> (this is what we are predicting)
   * shot_type
   * shot_zone_area
   * shot_zone_basic
   * shot_zone_range
   * team_id
   * team_name
   * game_date
   * matchup
   * opponent
   * shot_id


In [9]:
# Clean data
actiontypes = dict(df.action_type.value_counts())
df['type'] = df.apply(lambda row: row['action_type'] if actiontypes[row['action_type']] > 20\
                          else row['combined_shot_type'], axis=1)
df.drop(['action_type', 'combined_shot_type'], axis=1, inplace=True)

df['away'] = df.matchup.str.contains('@')
df.drop('matchup', axis=1, inplace=True)

df['distance'] = df.apply(lambda row: row['shot_distance'] if row['shot_distance'] <45 else 45, axis=1)

df['time_remaining'] = df.apply(lambda row: row['minutes_remaining'] * 60 + row['seconds_remaining'], axis=1)
df['last_moments'] = df.apply(lambda row: 1 if row['time_remaining'] < 3 else 0, axis=1)

data = pd.get_dummies(df['type'],prefix="action_type")

features=["away", "period", "playoffs", "shot_type", "shot_zone_area", "shot_zone_basic", "season",
           "shot_zone_range", "opponent", "distance", "minutes_remaining", "last_moments"]
for f in features:
    data = pd.concat([data, pd.get_dummies(df[f], prefix=f),], axis=1)

In [38]:
# Need work on game_date, add this into feature and increse n_estimators can inprove results but waste time and memory 
X = data[~mask]
y = df.shot_made_flag[~mask]

In [3]:
data
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 25697 entries, 19214 to 19185
Data columns (total 27 columns):
Unnamed: 0              25697 non-null int64
Unnamed: 0.1            25697 non-null int64
Unnamed: 0.1.1          25697 non-null int64
Unnamed: 0.1.1.1        25697 non-null int64
Unnamed: 0.1.1.1.1      25697 non-null int64
Unnamed: 0.1.1.1.1.1    25697 non-null int64
loc_x                   25697 non-null int64
loc_y                   25697 non-null int64
minutes_remaining       25697 non-null int64
period                  25697 non-null int64
playoffs                25697 non-null int64
season                  25697 non-null object
seconds_remaining       25697 non-null int64
shot_distance           25697 non-null int64
shot_made_flag          25697 non-null float64
shot_type               25697 non-null object
shot_zone_area          25697 non-null object
shot_zone_basic         25697 non-null object
shot_zone_range         25697 non-null object
game_date               2

## Initiating H2O server

In [3]:
import h2o
from h2o.automl import H2OAutoML
h2o.init()

Checking whether there is an H2O instance running at http://localhost:54321..... not found.
Attempting to start a local H2O server...
; OpenJDK 64-Bit Server VM (build 25.152-b12, mixed mode)56-b12)
  Starting server from C:\Users\prabh\Anaconda3\lib\site-packages\h2o\backend\bin\h2o.jar
  Ice root: C:\Users\prabh\AppData\Local\Temp\tmphypnslfl
  JVM stdout: C:\Users\prabh\AppData\Local\Temp\tmphypnslfl\h2o_prabh_started_from_python.out
  JVM stderr: C:\Users\prabh\AppData\Local\Temp\tmphypnslfl\h2o_prabh_started_from_python.err
  Server is running at http://127.0.0.1:54321
Connecting to H2O server at http://127.0.0.1:54321... successful.


--------------------------  --------------------------------
H2O cluster uptime:         04 secs
H2O cluster version:        3.12.0.1
H2O cluster version age:    1 year, 8 months and 20 days !!!
H2O cluster name:           H2O_from_python_prabh_8sk2ta
H2O cluster total nodes:    1
H2O cluster free memory:    3.512 Gb
H2O cluster total cores:    12
H2O cluster allowed cores:  12
H2O cluster status:         accepting new members, healthy
H2O connection url:         http://127.0.0.1:54321
H2O connection proxy:
H2O internal security:      False
Python version:             3.7.2 final
--------------------------  --------------------------------

#### Loading the data to the server cleansed

In [14]:
# Import a sample binary outcome train/test set into H2O
train = h2o.import_file("data.csv")

Parse progress: |█████████████████████████████████████████████████████████| 100%


##### Was trying to check the correlation of each columns with others for understanding the data

In [17]:
data.corr()

Unnamed: 0  Unnamed: 0.1  Unnamed: 0.1.1  \
Unnamed: 0                1.000000      1.000000        1.000000   
Unnamed: 0.1              1.000000      1.000000        1.000000   
Unnamed: 0.1.1            1.000000      1.000000        1.000000   
Unnamed: 0.1.1.1          1.000000      1.000000        1.000000   
Unnamed: 0.1.1.1.1        1.000000      1.000000        1.000000   
Unnamed: 0.1.1.1.1.1      1.000000      1.000000        1.000000   
Unnamed: 0.1.1.1.1.1.1    0.999997      0.999997        0.999997   
game_event_id             0.025521      0.025521        0.025521   
game_id                   0.760329      0.760329        0.760329   
lat                      -0.033551     -0.033551       -0.033551   
loc_x                    -0.012451     -0.012451       -0.012451   
loc_y                     0.033551      0.033551        0.033551   
lon                      -0.012451     -0.012451       -0.012451   
minutes_remaining        -0.008314     -0.008314       -0.008314   
period                   -0.003366     -0.003366       -0.003366   
playoffs                  0.612019      0.612019        0.612019   
seconds_remaining        -0.004856     -0.004856       -0.004856   
shot_distance             0.020567      0.020567        0.020567   
shot_made_flag           -0.013070     -0.013070       -0.013070   
team_id                        NaN           NaN             NaN   
shot_id                   0.999997      0.999997        0.999997   

                        Unnamed: 0.1.1.1  Unnamed: 0.1.1.1.1  \
Unnamed: 0                      1.000000            1.000000   
Unnamed: 0.1                    1.000000            1.000000   
Unnamed: 0.1.1                  1.000000            1.000000   
Unnamed: 0.1.1.1                1.000000            1.000000   
Unnamed: 0.1.1.1.1              1.000000            1.000000   
Unnamed: 0.1.1.1.1.1            1.000000            1.000000   
Unnamed: 0.1.1.1.1.1.1          0.999997            0.999997   
game_event_id                   0.025521            0.025521   
game_id                         0.760329            0.760329   
lat                            -0.033551           -0.033551   
loc_x                          -0.012451           -0.012451   
loc_y                           0.033551            0.033551   
lon                            -0.012451           -0.012451   
minutes_remaining              -0.008314           -0.008314   
period                         -0.003366           -0.003366   
playoffs                        0.612019            0.612019   
seconds_remaining              -0.004856           -0.004856   
shot_distance                   0.020567            0.020567   
shot_made_flag                 -0.013070           -0.013070   
team_id                              NaN                 NaN   
shot_id                         0.999997            0.999997   

                        Unnamed: 0.1.1.1.1.1  Unnamed: 0.1.1.1.1.1.1  \
Unnamed: 0                          1.000000                0.999997   
Unnamed: 0.1                        1.000000                0.999997   
Unnamed: 0.1.1                      1.000000                0.999997   
Unnamed: 0.1.1.1                    1.000000                0.999997   
Unnamed: 0.1.1.1.1                  1.000000                0.999997   
Unnamed: 0.1.1.1.1.1                1.000000                0.999997   
Unnamed: 0.1.1.1.1.1.1              0.999997                1.000000   
game_event_id                       0.025521                0.025433   
game_id                             0.760329                0.761232   
lat                                -0.033551               -0.033476   
loc_x                              -0.012451               -0.012453   
loc_y                               0.033551                0.033476   
lon                                -0.012451               -0.012453   
minutes_remaining                  -0.008314               -0.008251   
period                             -0.

##### We signify the target variable for the H2O

In [17]:
# Identify predictors and response
x = train.columns
y = "shot_made_flag"
x.remove(y)

##### And then convert the column to a factor to make it easy for the H2O. Since H2O considers the data type of the variables and predictors on its own, which might turn out to be irrelevant and unfavourable for the model generation. We make sure by converting it manually and make the H2O interpret better during the feature selection.

In [18]:
# For binary classification, response should be a factor
train[y] = train[y].asfactor()

Before we start training H2O with our data, we tried in many simple methods first by not denoting few parameters to it. 
* First, we tried giving no run_time to it, so by `DEFAULT` it consideres the run_time for 1 hour.
While we dont want it to run that long, even though it gave high score(Log Loss) at the end. In the next step we signified few parameters to it to facilitate
* Second, we ran H2O for around 1000 senconds and came up with a better score than the First - `0.599577`

<img src="Images/Capture.PNG"/>

* After a fine tweaking and managing with the parameters, we specified the `seed="1"` and `run_time=1000sec` and ran the H2O as shown below.

In [22]:
# Run AutoML for 20 base models (limited to 1 hour max runtime by default)
# aml = H2OAutoML(max_models=20, seed=1, max_runtime_secs=500)
aml = H2OAutoML(max_runtime_secs=1000, seed=1, max_models=100)
aml.train(x=x, y=y, training_frame=train)

AutoML progress: |████████████████████████████████████████████████████████| 100%
Parse progress: |█████████████████████████████████████████████████████████| 100%


In [23]:
# View the AutoML Leaderboard
lb = aml.leaderboard
lb.head(rows=lb.nrows)  # Print all rows instead of default (10 rows)

model_id auc logloss 
 
 
 DRF_0_AutoML_20190227_183234 0.948715 0.344336 
 XRT_0_AutoML_20190227_183234 0.923095 0.484171 
 StackedEnsemble_0_AutoML_20190227_183234 0.795427 0.562137 
 GBM_grid_0_AutoML_20190227_183234_model_0 0.785895 0.621769 
 GLM_grid_0_AutoML_20190227_183234_model_0 0.711312 0.600094 
 GLM_grid_0_AutoML_20190227_185017_model_1 0.702603 0.603857 
 GBM_grid_0_AutoML_20190227_185017_model_3 0.701112 0.680623 
 GLM_grid_0_AutoML_20190227_183234_model_1 0.699046 0.608849 
 GBM_grid_0_AutoML_20190227_185017_model_2 0.697811 0.619767 
 GLM_grid_0_AutoML_20190227_185017_model_0 0.69258 0.612955

##### At the end, we got pretty good `Log Loss = 0.344336` which is a very good score

To show how our kernel would rank in the Kaggle competition :


<img src="Images/image1.jpg" width="700" >

##### Now, for further analysis of the model that are generated, we consider the best model generated

H2O gives you the best features that are to be considered for the best model automatically. And we can check the same.


<img src="Images/image2.PNG" width="500" />

#### and from this we see, that the important variables turned out to be :

* variable	
* matchup
* action_type	
* opponent	
* seconds_rema
* game_event_id	
* Unnamed: 0.1.1.1.1
* shot_zone_range
* game_id
* shot_type
* playoffs


In [26]:
aml_leaderboard_df=aml.leaderboard.as_data_frame()
model_set=aml_leaderboard_df['model_id']
mod_best=h2o.get_model(model_set[0])

In [34]:
print(mod_best.logloss())
print(mod_best.algo)
print(mod_best.model_id)
print(mod_best.gini)

0.6922279011323961
drf
DRF_0_AutoML_20190227_183234
Model Details
H2ORandomForestEstimator :  Distributed Random Forest
Model Key:  DRF_0_AutoML_20190227_183234


ModelMetricsBinomial: drf
** Reported on train data. **

MSE: 0.22555366606423846
RMSE: 0.474924905710617
LogLoss: 0.6922279011323961
Mean Per-Class Error: 0.3614601725940524
AUC: 0.6760012086173288
Gini: 0.35200241723465764
Confusion Matrix (Act/Pred) for max f1 @ threshold = 0.2237670104211395: 


0     1      Error    Rate
-----  ----  -----  -------  ----------------
0      2042  7840   0.7934   (7840.0/9882.0)
1      852   7156   0.1064   (852.0/8008.0)
Total  2894  14996  0.4859   (8692.0/17890.0)

Maximum Metrics: Maximum metrics at their respective thresholds



metric                       threshold    value     idx
---------------------------  -----------  --------  -----
max f1                       0.223767     0.622153  323
max f2                       0.00849066   0.802672  398
max f0point5                 0.562234     0.612044  145
max accuracy                 0.578689     0.656177  138
max precision                1            0.944606  0
max recall                   2.17486e-05  1         399
max specificity              1            0.998077  0
max absolute_mcc             0.663614     0.303363  106
max min_per_class_accuracy   0.421242     0.621534  214
max mean_per_class_accuracy  0.500013     0.63854   174

Gains/Lift Table: Avg response rate: 44.76 %



group    cumulative_data_fraction    lower_threshold    lift     cumulative_lift    response_rate    cumulative_response_rate    capture_rate    cumulative_capture_rate    gain     cumulative_gain
--  -------  --------------------------  -----------------  -------  -----------------  ---------------  --------------------------  --------------  -------------------------  -------  -----------------
    1        0.0140861                   1                  2.23402  2.23402            1                1                           0.0314685       0.0314685                  123.402  123.402
    2        0.0200112                   0.986674           2.23402  2.23402            1                1                           0.0132368       0.0447053                  123.402  123.402
    3        0.032588                    0.98               2.23402  2.23402            1                1                           0.0280969       0.0728022                  123.402  123.402
    4        0.0400224                   0.968334           2.23402  2.23402            1                1                           0.0166084       0.0894106                  123.402  123.402
    5        0.0528787                   0.96               2.23402  2.23402            1                1                           0.0287213       0.118132                   123.402  123.402
    6        0.1                         0.913707           2.23402  2.23402            1                1                           0.10527         0.223402                   123.402  123.402
    7        0.150028                    0.861647           2.23402  2.23402            1                1                           0.111763        0.335165                   123.402  123.402
    8        0.2                         0.816164           2.23402  2.23402            1                1                           0.111638        0.446803                   123.402  123.402
    9        0.3                         0.730743           2.23402  2.23402            1                1                           0.223402        0.670205                   123.402  123.402
    10       0.4                         0.641427           2.23402  2.23402            1                1                           0.223402        0.893606                   123.402  123.402
    11       0.5                         0.282262           1.06394  2                  0.476244         0.895249                    0.106394        1                          6.39361  100
    12       0.6                         0.209605           0        1.66667            0                0.746041                    0               1                          -100     66.6667
    13       0.7                         0.166995           0        1.42857            0                0.639463                    0               1                          -100     42.8571
    14       0.800056                    0.130667           0        1.24991            0                0.559491                    0               1                          -100     24.9913
    15       0.9                         0.0928255          0        1.11111            0                0.49736                     0               1                          -100     11.1111
    16       1                           0                  0        1                  0                0.447624                    0               1                          -100     0



ModelMetricsBinomial: drf
** Reported on validation data. **

MSE: 0.22106351697169108
RMSE: 0.4701739220455459
LogLoss: 0.6504722093837271
Mean Per-Class Error: 0.3593506824850108
AUC: 0.6764376703609113
Gini: 0.3528753407218226
Confusion Matrix (Act/Pred) for max f1 @ threshold = 0.253813820765132: 


0    1     Error    Rate
-----  ---  ----  -------  ---------------
0      447  1758  0.7973   (1758.0/2205.0)
1      184  1558  0.1056   (184.0/1742.0)
Total  631  3316  0.492    (1942.0/3947.0)

Maximum Metrics: Maximum metrics at their respective thresholds



metric                       threshold    value     idx
---------------------------  -----------  --------  -----
max f1                       0.253814     0.616054  316
max f2                       0            0.797984  399
max f0point5                 0.553561     0.617895  150
max accuracy                 0.553561     0.664555  150
max precision                0.972533     0.967213  6
max recall                   0            1         399
max specificity              0.999927     0.999546  0
max absolute_mcc             0.679991     0.321287  99
max min_per_class_accuracy   0.419629     0.622273  218
max mean_per_class_accuracy  0.553561     0.640649  150

Gains/Lift Table: Avg response rate: 44.13 %



group    cumulative_data_fraction    lower_threshold    lift      cumulative_lift    response_rate    cumulative_response_rate    capture_rate    cumulative_capture_rate    gain      cumulative_gain
--  -------  --------------------------  -----------------  --------  -----------------  ---------------  --------------------------  --------------  -------------------------  --------  -----------------
    1        0.0134279                   0.98               2.18029   2.18029            0.962264         0.962264                    0.0292767       0.0292767                  118.029   118.029
    2        0.0215353                   0.96               2.19498   2.18582            0.96875          0.964706                    0.0177956       0.0470723                  119.498   118.582
    3        0.0301495                   0.922231           1.99922   2.1325             0.882353         0.941176                    0.0172216       0.0642939                  99.9223   113.25
    4        0.0400304                   0.900698           1.85911   2.06502            0.820513         0.911392                    0.0183697       0.0826636                  85.9107   106.502
    5        0.0506714                   0.88               1.99605   2.05054            0.880952         0.905                       0.02124         0.103904                   99.605    105.054
    6        0.100076                    0.782724           1.74291   1.89867            0.769231         0.837975                    0.0861079       0.190011                   74.2913   89.8672
    7        0.149987                    0.705475           1.70222   1.8333             0.751269         0.809122                    0.0849598       0.274971                   70.2215   83.3297
    8        0.200659                    0.64               1.31416   1.7022             0.58             0.751263                    0.0665901       0.341561                   31.4156   70.2201
    9        0.299975                    0.531777           1.15023   1.51945            0.507653         0.670608                    0.114237        0.455798                   15.0233   51.9455
    10       0.400051                    0.461897           0.906315  1.36607            0.4              0.602913                    0.0907003       0.546498                   -9.36854  36.6073
    11       0.500127                    0.410042           0.877634  1.26834            0.387342         0.559777                    0.0878301       0.634328                   -12.2366  26.8335
    12       0.599949                    0.364931           0.862609  1.20083            0.380711         0.529983                    0.0861079       0.720436                   -13.7391  20.0829
    13       0.700025                    0.322281           0.705549  1.13002            0.311392         0.498733                    0.0706085       0.791045                   -29.4451  13.0023
    14       0.799848                    0.274319           0.730342  1.08014            0.322335         0.476718                    0.0729047       0.863949                   -26.9658  8.01421
    15       0.899924                    0.216979           0.653923  1.03274            0.288608         0.4558                      0.065442        0.929392                   -34.6077  3.27444
    16       1                           0                  0.705549  1                  0.311392         0.441348                    0.0706085       1                          -29.4451  0



ModelMetricsBinomial: drf
** Reported on cross-validation data. **

MSE: 0.218546364440412
RMSE: 0.46748942708943914
LogLoss: 0.631739907321219
Mean Per-Class Error: 0.35333141104999033
AUC: 0.6858717519578176
Gini: 0.37174350391563515
Confusion Matrix (Act/Pred) for max f1 @ threshold = 0.24677690550684933: 


0     1      Error    Rate
-----  ----  -----  -------  ----------------
0      2090  7792   0.7885   (7792.0/9882.0)
1      794   7214   0.0992   (794.0/8008.0)
Total  2884  15006  0.4799   (8586.0/17890.0)

Maximum Metrics: Maximum metrics at their respective thresholds



metric                       threshold    value     idx
---------------------------  -----------  --------  -----
max f1                       0.246777     0.626923  324
max f2                       0.0479236    0.802431  395
max f0point5                 0.581718     0.63367   139
max accuracy                 0.581718     0.669424  139
max precision                0.999979     0.975309  0
max recall                   0.0199886    1         398
max specificity              0.999979     0.999798  0
max absolute_mcc             0.630246     0.333827  118
max min_per_class_accuracy   0.418678     0.626998  223
max mean_per_class_accuracy  0.564413     0.646669  148

Gains/Lift Table: Avg response rate: 44.76 %



group    cumulative_data_fraction    lower_threshold    lift      cumulative_lift    response_rate    cumulative_response_rate    capture_rate    cumulative_capture_rate    gain      cumulative_gain
--  -------  --------------------------  -----------------  --------  -----------------  ---------------  --------------------------  --------------  -------------------------  --------  -----------------
    1        0.0124651                   0.98               2.13384   2.13384            0.955157         0.955157                    0.0265984       0.0265984                  113.384   113.384
    2        0.0218558                   0.96               2.14093   2.13688            0.958333         0.956522                    0.0201049       0.0467033                  114.093   113.688
    3        0.030408                    0.94               1.98579   2.09439            0.888889         0.9375                      0.016983        0.0636863                  98.5792   109.439
    4        0.0405813                   0.91               1.97624   2.06477            0.884615         0.924242                    0.0201049       0.0837912                  97.6245   106.477
    5        0.0500279                   0.892537           1.9432    2.04182            0.869822         0.913966                    0.0183566       0.102148                   94.3197   104.182
    6        0.1                         0.804055           1.77672   1.90934            0.795302         0.854667                    0.0887862       0.190934                   77.6717   90.9341
    7        0.150028                    0.727868           1.64993   1.82284            0.738547         0.815946                    0.0825425       0.273477                   64.9927   82.2837
    8        0.2                         0.654465           1.47185   1.73514            0.658837         0.776691                    0.0735514       0.347028                   47.1852   73.514
    9        0.300727                    0.54               1.12941   1.53225            0.505549         0.685874                    0.113761        0.460789                   12.9405   53.2253
    10       0.4                         0.465719           0.919519  1.38018            0.411599         0.617803                    0.0912837       0.552073                   -8.0481   38.0182
    11       0.5                         0.410486           0.869131  1.27797            0.389044         0.572051                    0.0869131       0.638986                   -13.0869  27.7972
    12       0.6                         0.363146           0.782967  1.19547            0.350475         0.535122                    0.0782967       0.717283                   -21.7033  19.5471
    13       0.7                         0.317046           0.757992  1.13297            0.339296         0.507147                    0.0757992       0.793082                   -24.2008  13.2974
    14       0.8                         0.267745           0.765485  1.08704            0.34265          0.486585                    0.0765485       0.86963                    -23.4515  8.7038
    15       0.9                         0.208283           0.725524  1.04687            0.324762         0.468604                    0.0725524       0.942183                   -27.4476  4.68698
    16       1                           0                  0.578172  1                  0.258804         0.447624                    0.0578172       1                          -42.1828  0


Cross-Validation Metrics Summary: 


mean      sd          cv_1_valid    cv_2_valid    cv_3_valid    cv_4_valid    cv_5_valid
-----------------------  --------  ----------  ------------  ------------  ------------  ------------  ------------
accuracy                 0.510397  0.0170002   0.46758       0.518726      0.529346      0.534097      0.502236
auc                      0.686009  0.00484444  0.683449      0.682677      0.676458      0.69192       0.695542
err                      0.489603  0.0170002   0.53242       0.481274      0.470654      0.465903      0.497764
err_count                1751.8    60.8266     1905          1722          1684          1667          1781
f0point5                 0.527635  0.00874553  0.504351      0.534937      0.536315      0.53709       0.52548
f1                       0.628575  0.00452619  0.616314      0.63517       0.630053      0.630459      0.630881
f2                       0.777918  0.00877781  0.792172      0.781625      0.763497      0.763121      0.789173
lift_top_group           2.13415   0.0233789   2.17334       2.10871       2.16781       2.0878        2.13309
logloss                  0.63174   0.00694624  0.647016      0.627896      0.638697      0.62545       0.619641
max_per_class_error      0.825074  0.0456959   0.928997      0.817391      0.765903      0.75253       0.860548
mcc                      0.148847  0.0142839   0.11269       0.155004      0.159952      0.172317      0.144271
mean_per_class_accuracy  0.550085  0.0106596   0.524632      0.553103      0.561562      0.567555      0.543574
mean_per_class_error     0.449915  0.0106596   0.475368      0.446897      0.438438      0.432445      0.456426
mse                      0.218546  0.00183199  0.219034      0.218781      0.222884      0.216923      0.215109
precision                0.47669   0.0103033   0.449868      0.484017      0.487921      0.488828      0.472818
r2                       0.115999  0.0074753   0.109783      0.117274      0.0997495     0.122723      0.130465
recall                   0.925244  0.0245832   0.978261      0.923598      0.889027      0.887641      0.947696
rmse                     0.467481  0.00195702  0.468011      0.467741      0.472106      0.46575       0.463799
specificity              0.174926  0.0456959   0.071003      0.182609      0.234097      0.24747       0.139452

Scoring History: 


timestamp            duration    number_of_trees    training_rmse    training_logloss    training_auc    training_lift    training_classification_error    validation_rmse    validation_logloss    validation_auc    validation_lift    validation_classification_error
--  -------------------  ----------  -----------------  ---------------  ------------------  --------------  ---------------  -------------------------------  -----------------  --------------------  ----------------  -----------------  ---------------------------------
    2019-02-27 18:33:17  42.067 sec  0                  nan              nan                 nan             nan              nan                              nan                nan                   nan               nan                nan
    2019-02-27 18:33:17  42.201 sec  1                  0.641715         13.3249             0.5684          1.85631          0.544475                         0.625783           12.6196               0.586524          1.25407            0.558652
    2019-02-27 18:33:17  42.375 sec  2                  0.621071         11.6372             0.576706        2.15143          0.549944                         0.556123           6.29379               0.60375           1.45134            0.558652
    2019-02-27 18:33:17  42.537 sec  3                  0.605196         10.3192             0.584189        2.2115           0.552793                         0.527434           3.76493               0.621496          1.61965            0.558652
    2019-02-27 18:33:18  42.711 sec  4                  0.588785         8.88045             0.591611        2.22955          0.55243                          0.513245           2.47211               0.631047          1.74704            0.558652
    2019-02-27 18:33:18  42.884 sec  5                  0.577025         7.86245             0.597519        2.23313          0.551722                         0.504595           1.87038               0.639911          1.83421            0.558652
    2019-02-27 18:33:18  43.069 sec  6                  0.567041         6.98976             0.602004        2.23402          0.551429                         0.500586           1.55489               0.641257          1.88191            0.525462
    2019-02-27 18:33:18  43.321 sec  7                  0.554397         6.00153             0.61078         2.23402          0.552169                         0.496483           1.30447               0.645522          1.95682            0.53281
    2019-02-27 18:33:18  43.557 sec  8                  0.545126         5.12618             0.613878        2.23402          0.552574                         0.493584           1.1362                0.646453          1.968              0.539144
    2019-02-27 18:33:19  43.787 sec  9                  0.537579         4.45405             0.618545        2.23402          0.552223                         0.489372           0.971402              0.650338          2.0319             0.536863
    2019-02-27 18:33:19  44.045 sec  10                 0.530031         3.90124             0.625078        2.23402          0.552397                         0.487149           0.925917              0.6536            2.02644            0.53965
    2019-02-27 18:33:19  44.328 sec  11                 0.522983         3.35632             0.630873        2.23402          0.552407                         0.485919           0.85961               0.653975          2.0915             0.543451
    2019-02-27 18:33:20  44.600 sec  12                 0.517532         2.96102             0.635426        2.23402          0.552357                         0.484722           0.847542              0.655315          2.07213            0.545984
    2019-02-27 18:33:20  44.883 sec  13                 0.511472         2.55636             0.641683        2.23402          0.510837                         0.483024           0.81919               0.657888          2.05792            0.529516
    2019-02-27 18:33:20  45

Variable Importances: 


variable            relative_importance    scaled_importance      percentage
------------------  ---------------------  ---------------------  ---------------------
matchup             18858.021484375        1.0                    0.14551283536919715
action_type         17095.373046875        0.9065305743256015     0.131911834218985
opponent            10931.6328125          0.5796807910924013     0.08435099552091456
seconds_remaining   7407.83251953125       0.3928213002445184     0.05716054119199457
game_event_id       6554.1416015625        0.347551921445895      0.05057326552760466
---                 ---                    ---                    ---
Unnamed: 0.1.1.1.1  2066.57666015625       0.10958608048402811    0.015946181288227973
shot_zone_range     1420.807373046875      0.07534233504952251    0.010963276796392061
game_id             725.1743774414062      0.03845442524510096    0.005595612449907475
shot_type           384.2039794921875      0.020373504177547123   0.0029646063592810652
playoffs            98.1711654663086       0.0052058040949655975  0.0007575113142350215


See the whole table with table.as_data_frame()
<bound method ModelBase.gini of >


## Conclusion

The H2O helps us execute and run the dataset for different models for different hyperparameter values, and get the best out of it using the leaderboard it generates. We use this leaderboard to get the model and use it for further analysis. We can see, that by giving the appropriate parameters to H2O it can give us the best models with best metric value and make us win the Kaggle competition. Thus, signifying that even we students can win a Kaggle competition efficiently.
Thus, the leaderboard turned out to give us the below algorithms with the metric as shown below for the prediction of Kobe Bryants' shots:

<img src='Images/image3.PNG' width="500" />
    
##### And it turned out that DRF was the algorithm that gave us best model

## Contribution

Tried to dig through with the way H2O works. Trying to check what parameters can be tuned and given to it to make better models with best scores.
* Own - 80%
* Other sources - 20%

## Citations

Prof. GitHub H2O - https://github.com/nikbearbrown/CSYE_7245/tree/master/H2O

H2O - https://h2o.ai

Kaggle competition - https://www.kaggle.com/c/kobe-bryant-shot-selection

Kaggle kernels - 
* https://www.kaggle.com/selfishgene/psychology-of-a-professional-athlete/data
* https://www.kaggle.com/apapiu/exploring-kobe-s-shots


## License

<font size="4">MIT License</font>
    
<img src="Images/OSI_Approved_License.png" width="100" align="right"/>

<font size="4">
    
<b>Copyright (c) 2019 PRABHU SUBRAMANIAN, PREETAM JAIN</b>

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
